xchange-stream-ftx/src/main/java/info/bitrich/xchangestream/ftx/FtxStreamingAdapters.java

```diff
@@ -0,0 +1,94 @@
+import org.knowm.xchange.dto.trade.LimitOrder;
+import org.knowm.xchange.instrument.Instrument;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Date;
+import java.util.List;
+import java.util.zip.CRC32;
+
+public class FtxStreamingAdapters {
+
+    private final static ObjectMapper mapper = StreamingObjectMapperHelper.getObjectMapper();
+
+    public static OrderBook adaptOrderbookMessage(OrderBook orderBook, Instrument instrument, JsonNode jsonNode){
+
+        Streams.stream(jsonNode)
+                .filter(JsonNode::isObject)
+                .map(res->{
+                    try{
+                        return mapper.readValue(res.toString(),FtxOrderbookResponse.class);
+                    }catch (IOException e){
+                        throw new RuntimeException(e);
+                    }
+                })
+                .forEach(
+                        message -> {
+
+                            if(message.getAction().equals("partial")){
+                                message.getAsks().forEach(ask->{
+                                    orderBook.getAsks().add(new LimitOrder.Builder(Order.OrderType.ASK,instrument).limitPrice(ask.get(0)).originalAmount(ask.get(1)).build());
+                                });
+
+                                message.getBids().forEach(bid ->{
+                                    orderBook.getBids().add(new LimitOrder.Builder(Order.OrderType.BID,instrument).limitPrice(bid.get(0)).originalAmount(bid.get(1)).build());
+                                });
+                            }else{
+                                message.getAsks().forEach(ask->
+                                        orderBook.update(new LimitOrder.Builder(Order.OrderType.ASK,instrument).limitPrice(ask.get(0)).originalAmount(ask.get(1)).build())
+                                );
+                                message.getBids().forEach(bid->
+                                        orderBook.update(new LimitOrder.Builder(Order.OrderType.BID,instrument).limitPrice(bid.get(0)).originalAmount(bid.get(1)).build())
+                                );
+                            }
+
+                            if(orderBook.getAsks().size() > 0 && orderBook.getBids().size() > 0){
+                                Long calculatedChecksum = getOrderbookChecksum(orderBook.getAsks(),orderBook.getBids());
+
+                                if(!calculatedChecksum.equals(message.getChecksum())){
+                                    throw new RuntimeException("Checksum is not correct!");
+                                }
+                            }
+                        });
+
+        return new OrderBook(Date.from(Instant.now()), Lists.newArrayList(orderBook.getAsks()),Lists.newArrayList(orderBook.getBids()),true);
+    }
+
+    public static Long getOrderbookChecksum(List<LimitOrder> asks,List<LimitOrder> bids){
+        StringBuilder data = new StringBuilder();
```

Comment:
can you get an estimation of the string size when complete, and find an approximation and make the default size that, a single large allocation is more efficient then starting at 16 growing, allocating another array, gc'ing and copying over to new array etc.

Reply:
Do you mean the StringBuilder to have a fixed size?

Reply:
stringbuilder will never have a fixed size, it's just about its initial size, you want to have an initial size that it won't have to create a new backing array during the creation process, if you see the default starting size is 16

Reply:
you mean to do the following: 

StringBuilder data = new StringBuilder(16);

Reply:
No, that is what happens by default, 16 is the default size.

The way this works under the hood the stringbuilder is backed by an array of size 16, so if you say, sb.append("test") the array of 16 now has 4 spaces used, if you sb.append("111222333444555") it will not fit all those characters in this array since total qty is 19. 

What happens now? That array is thrown away and a new one of size 32 is allocated and all the contents are copied over. This means everytime the stringbuilder has enough contents to exceed the backing array's size, it 1) creates a new array of size 2*prev_size 2) copies the contents over 3) gc's the old contents.

This is an expensive thing to do if we know what the final array size or at least a better sense of the final size of the array will be so rather then do this you can just say something like new StringBuilder(2048) or a number that is close to the final size of the string and only allocate one time.

Does this make sense?

Reply:
Thank you very much for the explanation! The thing is that every currencyPair has different prices and even the volume of every currencyPair has different scale settings. So, i should either put an extreme number like 5000000 or just leave it as it is. The checksum gets the first 100 prices on both sides(ask and bid) and adds them separated by a colon. So, it is a very big String. What do you thing?

Reply:
You can probably do a reasonable estimation of size that will give you a good starting point:

100*(8 + 1 + 6)
8 = len("60000.00") (price)
1 = ":"
5 = len("12.01") (qty) = 2400

So to be slightly aggressive = 3072 works? if it uses less nbd it will get collected if it uses more same thing it will only require maybe one more copy

Reply:
Great, thanks!